### PR TITLE
Add federated learning lib code and example

### PR DESCRIPTION
**Pull request by kason-huang**
1. Add lib code for federated learning 
2. Add example for federated learning

### CODE REVIEW

File: examples/surface_defect_detection/README.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+Using Federated Learning, we can solve the problem. Each place uses its own data for model training ,uploads the weight to the cloud for aggregation, and obtains the aggregation result for model update.
+
+
+## Surface Defect Detection Experiment
+> Assume that there are two edge nodes (edge1 and edge2) and a cloud node. Data on the edge nodes cannot be migrated to the cloud due to privacy issues.
+> Base on this scenario, we will demonstrate the surface inspection.
+
+### Install Neptune
+
+Follow the [Neptune installation document](docs/setup/install.md) to install Neptune.
```
Comment (llhuii) [inline examples/surface_defect_detection/README.md:13]: `docs/setup/install.md` => `/docs/setup/install.md`

File: examples/surface_defect_detection/README.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+```
+# create dataset for edge1
+kubectl create -f - <<EOF
+apiVersion: neptune.io/v1alpha1
+kind: Dataset
+metadata:
+  name: "edge1-surface-defect-detection-dataset"
+spec:
+  url: "/data/1.txt"
+  format: "txt"
```
Comment (llhuii) [inline examples/surface_defect_detection/README.md:40]: NodeName field is missed

File: examples/surface_defect_detection/README.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+```
+
+### Check Federated Learning Status
+
+```
+kubectl get federatedlearningjob surface-defect-detection
+```
+
+### Check Federated Learning Train Result
+After the job completed, you will find the model generated on the path `/home/data/model` in edge1 and edge2.
```
Comment (llhuii) [inline examples/surface_defect_detection/README.md:131]: `/home/data/model`? It should be `/model` at `cloud0`.

File: examples/surface_defect_detection/README.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+
+
```
Comment (llhuii) [inline examples/surface_defect_detection/README.md:134]: 多余空行删除

File: lib/neptune/federated_learning/aggregator.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                ), self._ws_timeout)
+                await websocket.send(agg_data.to_json())
+                LOG.info(f"size of the message "
+                         f"is {len(agg_data.to_json())}")
+                result = await websocket.recv()
+                await websocket.close()
+                return result
+            except Exception as e:
+                error = e
+                LOG.error(f"send data error: {error}, type: {type(error)}")
```
Comment (llhuii) [inline lib/neptune/federated_learning/aggregator.py:187]: error改成warningf

Comment (llhuii) [reply lib/neptune/federated_learning/aggregator.py:187]: need to call websocket.close when exception.

File: lib/neptune/federated_learning/aggregator.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        error = None
+        for i in range(AggregationClient._retry):
+            try:
+                # TODO should not connect websocket each time, use long
+                # connection
+                websocket = await asyncio.wait_for(websockets.connect(
+                    self.ws_url,
+                    max_size=MAX_SIZE_BYTE
+                ), self._ws_timeout)
+                await websocket.send(agg_data.to_json())
```
Comment (llhuii) [inline lib/neptune/federated_learning/aggregator.py:179]: agg_data.to_json() should be cached in case too large.

Comment (llhuii) [reply lib/neptune/federated_learning/aggregator.py:179]: why to_json in that case websocket.send needs to serialize again?  why not just to_bytes?

File: lib/neptune/federated_learning/aggregator.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+                                    self.config.bind_ip,
+                                    self.config.bind_port,
+                                    max_size=MAX_SIZE_BYTE):
+            LOG.info("start websocket on "
+                     f"ws://{self.config.bind_ip}:{self.config.bind_port}")
+            await stop
+            LOG.info("websocket server close")
+
+    async def _receive(self, websocket, path):
+        """Receive the data form worker and aggregate, then send the result
```
Comment (llhuii) [inline lib/neptune/federated_learning/aggregator.py:113]: form typo

**Comment by llhuii** [comment]
suggest that second commit message say what's example

File: examples/surface_defect_detection/README.md (hunk 3 above)
Comment (khalid-davis) [reply examples/surface_defect_detection/README.md:131]: The cloud worker do not save the model result. Only the edge worker will save. It should be `/model` in edge1 and edge2

**Pull request merged**